### PR TITLE
Introduction to the test framework Test Cafe

### PR DESCRIPTION
**Pull request by jnorbert**

### CODE REVIEW

**Comment by pavelloz** [comment]
As a bonus we can link to our public repos where we use TC:
https://github.com/mdyd-dev/nearme-documentation/tree/master/tests/e2e
https://github.com/mdyd-dev/marketplace-nearme-example/tree/release_candidate/tests/e2e

It can help showcase more complete test suites and file organization.

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 1]
```diff
@@ -0,0 +1,148 @@
+---
+slug: tutorials/qa/testcafe
+metadata:
+  title: Introduction to the test framework Test Cafe
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:4]:
```suggestion
  title: Introduction to the test framework TestCafe
```

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+{% raw %}
+fixture `My First Fixture`
+    .page `https://examples.platform-os.com/full-form-example`;
+{% endraw %}
+```
+
+The whole test will look as following
+
+```liquid
+{% raw %}
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:44]: No need to use raw - there is no liquid in here, only js :)

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+<pre class="command-line" data-output="2-4" data-user="user" data-host="host"><code class="language-bash">
+npm install -g testcafe
+</code></pre>
+
+## Writing a Test
+To write the first test, we will use a site with examples from the platformOS documentation.
+
+First, we will determine the fixture declaration. A fixture is a test category and each test file requires its determination.
+
+```liquid
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:34]:
```suggestion
```js
```

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+{% raw %}
+fixture `My First Fixture`
+    .page `https://examples.platform-os.com/full-form-example`;
+{% endraw %}
+```
+
+The whole test will look as following
+
+```liquid
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:43]:
```suggestion
```js
```

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+<img src="{{ 'images/tutorials/qa/list_browsers.png' | asset_url }}" alt="List browsers" />
+
+## Page Object Model
+A good practice is to use the Page Object Model (POM) pattern. It consists of selectors, URL addresses and constants, which we will use in a separate "page" file. We will import this file to our test, and then assign the action to be executed on each element.
+Thanks to this approach, it will be easier to maintain the tests, because if there is a necessity to make some changes, the test itself will remain intact, and it will be possible to modify the selectors or URL addresses alone in one "page" file.
+
+In the catalog with an example test, create a /pages folder, and in it, a page-object.js file together with a Page class.
+
+```liquid
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:93]:
```suggestion
```js
```

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+{% raw %}
+export default class Page {
+    constructor () {
+    }
+}
+{% endraw %}
+```
+Next thing will be to add three selectors used in the test in the Page class.
+
+```liquid
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:103]:
```suggestion
```js
```

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    this.submitButton = Selector('button.btn.btn-primary');
+  }
+}
+{% endraw %}
+```
+
+The page-object.js link has to be created in the test file, as well as the instance of the Page class. 
+Then, we will use these fields in the test actions:
+
+```liquid
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:122]:
```suggestion
```js
```

File: marketplace_builder/views/pages/tutorials/qa/testcafe.liquid [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    .expect(page.successNotification.innerText).eql('This is flash notice (success)');
+});
+{% endraw %}
+```
+
+Let's start the test again and check its result.
+
+<img src="{{ 'images/tutorials/qa/test_results.png' | asset_url }}" alt="Test results" />
+
+I hope that the above article explained the possibilities of writing tests using Test Cafe.
```
Comment (pavelloz) [inline marketplace_builder/views/pages/tutorials/qa/testcafe.liquid:145]:
```suggestion
I hope that the above article explained the possibilities of writing tests using TestCafe. 
```

**Comment by pavelloz** [comment]
+ couple notes about formatting ;)

**Comment by diana-lakatos** [comment]
I'm working on this, breaking it down into separate topics, reviewing, editing, and fixing formatting based on Pawel's comments.

**Comment by diana-lakatos** [comment]
Reworked into 4 topics here: #447

**Pull request closed without merging**